### PR TITLE
Removed call to `TryCreateDirectory` from `GetDefaultDataDir` in `src/util.cpp`.

### PR DESCRIPTION
**Pull request by alexreg**
See https://github.com/bitcoin/bitcoin/issues/7845#issuecomment-207684728.

### CODE REVIEW

**Comment by kirkalx** [comment]
utACK. If there **is** a good reason for that line to be there on OSX only, then it needs a comment justifying it.

**Comment by sipa** [comment]
It seems that this was there as long as GetDefaultDatadir has been there: https://github.com/bitcoin/bitcoin/tree/d882773789ea3894de7163f7bb880c5b23072882/util.cpp

**Comment by paveljanik** [comment]
Looks like this is not for merging...

**Comment by alexreg** [comment]
??

**Comment by paveljanik** [comment]
According to the comment mentioned above, you are only trying if this will work. Your real problem is different though, I think. Can you investigate it a bit more please?

```
pavel$ sudo -u test env | grep HOME
HOME=/Users/pavel
pavel$ sudo -H -u test env | grep HOME
HOME=/Users/test
pavel$ 
```

**Comment by mrCertified** [comment]
pavel$ sudo -u test env | grep changes to semi HOME
HOME=/Users/pavel
pavel$ sudo -H -u test env | grep changes to semi HOME
HOME=/Users/test
pavel$ sudo -H -u final test

File: src/util.cpp [hunk 1]
```diff
@@ -472,7 +472,6 @@ boost::filesystem::path GetDefaultDataDir()
 #ifdef MAC_OSX
     // Mac
     pathRet /= "Library/Application Support";
-    TryCreateDirectory(pathRet);
     return pathRet / "Bitcoin";
```
Comment (mrCertified) [inline src/util.cpp]: TryCreateDirectory(pathRet);release pathRet / "Bitcoin";

File: src/util.cpp [hunk 2]
```diff
@@ -472,7 +472,6 @@ boost::filesystem::path GetDefaultDataDir()
 #ifdef MAC_OSX
     // Mac
     pathRet /= "Library/Application Support";
-    TryCreateDirectory(pathRet);
```
Comment (mrCertified) [inline src/util.cpp]: TryCreateDirectory(pathRet);

File: src/util.cpp (hunk 1 above)
Comment (mrCertified) [inline src/util.cpp]: return pathRet / "Bitcoin";

**Comment by paveljanik** [comment]
@mrCertified ?

**Comment by kirkalx** [comment]
@paveljanik The change is for merging. As I said above, if there is a reason for the line to be there it needs to be justified with a comment.

**Comment by sipa** [comment]
And I added that the behaviour has been there for as long as the
GetDefaultDatadir function has existed, implying that we'll probably never
find out why.

Concept ACK

**Comment by paveljanik** [comment]
@sipa yes, but someone should really investigate, how `bitcoind`/-Qt works when the parent directory (ie. `Library/Application Support`) is not created/is not writable by the current user or does not exist at all (where/how will we write the config file, blocks, chainstate etc?). I suspect this hides some other problem down in the code...

But if proven to work (not just compile ;-), I'm happy with making OS X the same as "Unix" branch there - @alexreg can you please make the OS X branch to directly return the full path and not changing it in two passes (1. add L/A S, 2. add /Bitcoin) and then test with a real user without that directory or with bad permissions on that directory? How will `bitcoind`/-Qt work in such cases?

**Comment by kirkalx** [comment]
But if the data directory has been changed to a different value, the only reason GetDefaultDataDir() is called is to display what the default would have been. (I'm ignoring the slightly more complicated behaviour in -qt). So making sure that (part of) that path exists makes no sense.

**Comment by paveljanik** [comment]
... exist and is writable...

**Comment by jonasschnelli** [comment]
utACK.
Creating `~/Library/Application Support` is useless.

**Comment by alexreg** [comment]
I’m not sure exactly what you mean. I don’t really code C++. Could you give me the diff please, and I’ll apply it?

I’ve already tested both those cases though, and it works as it should.

**Comment by paveljanik** [comment]
Just `return pathRet / "Library/Application Support/Bitcoin"` ...

**Comment by alexreg** [comment]
Oh right. Done now.

**Comment by sipa** [comment]
utACK 0f2fd2cec05b90d071b30eb522061e1c5144b367 after squash

**Comment by paveljanik** [comment]
Please squash both commits into one.

**Comment by alexreg** [comment]
I don’t use Git much, so no idea how to do that.

**Comment by kirkalx** [comment]
Wladimir explains in #7458, or just do a google search

**Comment by fanquake** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/7850/commits/41dbc4849e0bf14eef98962b0f0bddcde0bb3014

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/7850/commits/41dbc4849e0bf14eef98962b0f0bddcde0bb3014

**Comment by laanwj** [comment]
Hmm this is indeed weird, why would this need a TryCreateDirectory on OSX.

Happy that this solves your problem.

utACK https://github.com/bitcoin/bitcoin/pull/7850/commits/41dbc4849e0bf14eef98962b0f0bddcde0bb3014

**Comment by paveljanik** [comment]
I digged in the git history, but could not find something important. But here is the wild guess: on OS X, the config files are in `~/Library/Application Support`. On other unixes, config directory `.bitcoin` is directly in the home directory. And the wild guess is that this "mkdir -p parent_directory" had to be somewhere, so someone decided it could be here 8)

**Comment by alexreg** [comment]
Thanks for accepting this, guys.

**Comment by MarcoFalke** [comment]
Backported as part of #7938. Removing label 'Needs backport'.

**Pull request merged**